libraries/protocol/custom_authorities/restriction_predicate.hxx

```diff
@@ -0,0 +1,595 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <graphene/protocol/restriction_predicate.hpp>
+
+#include <fc/exception/exception.hpp>
+
+#include "safe_compare.hpp"
+
+namespace graphene { namespace protocol {
+namespace typelist = fc::typelist;
+using std::declval;
+using std::size_t;
+using restriction_function = restriction::function_type;
+using restriction_argument = restriction::argument_type;
+
+// Make our own std::void_t since the real one isn't available in C++14
+template<typename...> using make_void = void;
+
+// Metafunction to check if type is some instantiation of fc::safe
+template<typename> constexpr static bool is_safe = false;
+template<typename I> constexpr static bool is_safe<fc::safe<I>> = true;
+
+// Metafunction to check if type is a flat_set of any element type
+template<typename> struct is_flat_set_impl : std::false_type {};
+template<typename T> struct is_flat_set_impl<flat_set<T>> : std::true_type {};
+template<typename T> constexpr static bool is_flat_set = is_flat_set_impl<T>::value;
+
+// We use our own is_integral which does not consider bools integral (to disallow comparison between bool and ints)
+template<typename T> constexpr static bool is_integral = !std::is_same<T, bool>::value &&
+                                                         !std::is_same<T, safe<bool>>::value &&
+                                                         (is_safe<T> || std::is_integral<T>::value);
+
+// Metafunction to check if two types are comparable, which means not void_t, and either the same or both integral
+template<typename T, typename U>
+constexpr static bool comparable_types = !std::is_same<T, void_t>::value &&
+                                         (std::is_same<T, U>::value || (is_integral<T> && is_integral<U>));
+
+// Metafunction to check if type is a container
+template<typename, typename = void>
+struct is_container_impl : std::false_type {};
+template<typename T>
+struct is_container_impl<T, make_void<typename T::value_type, decltype(declval<T>().size())>> : std::true_type {};
+template<typename T> constexpr static bool is_container = is_container_impl<T>::value;
+
+// Type alias for a predicate on a particular field type
+template<typename Field>
+using object_restriction_predicate = std::function<predicate_result(const Field&)>;
+
+// Get the actual number when type might be a safe<I>
+template<typename I, typename=std::enable_if_t<std::is_integral<I>::value>>
+const auto& to_num(const I& i) { return i; }
+template<typename I>
+const auto& to_num(const fc::safe<I>& i) { return i.value; }
+inline auto to_num(const fc::time_point_sec& t) { return t.sec_since_epoch(); }
```

Comment:
Might deal with `unsigned_int` similarly here. Actually, IMHO `unsigned_int` should be treated as integral.

Reply:
See 3cd47d1